crates/ra_hir/src/ty.rs

```diff
@@ -422,7 +422,7 @@ impl Ty {
         }
     }
 
-    fn callable_sig(&self, db: &impl HirDatabase) -> Option<FnSig> {
+    pub fn callable_sig(&self, db: &impl HirDatabase) -> Option<FnSig> {
```

Comment:
Is there an alternative way of getting a function signature that doesn't require making this function public?

Reply:
I feel the approach of basically replicating a (very small) part of the type checking in the validation is a bit too ad-hoc anyway. Long-term, we would want a 'type mismatch' diagnostic produced by the type checker, and then have the Ok-wrapping produced as a quick-fix for that. So it might be nice to start building a simple version of that: See the `unify` call [here](https://github.com/rust-analyzer/rust-analyzer/blob/e3f8e6023d66ceb37040833bf317d77019bb10f1/crates/ra_hir/src/ty/infer.rs#L1309); this is the place where we combine the expected type of an expression (in the case of the function tail expression, the annotated return type) and the actual type. Note that `unify` returns a `bool` which should specify whether the types unified or not, but we don't use it. I'd propose that for now, if the unification returns false, we record that fact in another map (e.g. `type_mismatches`) in the `InferenceResult`, together with the expected type `expected.ty`. (We don't want to create an actual diagnostic for that yet, since there would be far too many false positives.)

Then you should be able to take a look at that in the validation and just use the actual and expected type to produce the fix.

Reply:
Thanks, I'll give that a go.

Reply:
I've implemented your suggestion. It was a suspiciously easy change!

Reply:
I *love* `type_mismatches` approach!